tox.ini

```diff
@@ -1,76 +1,124 @@
 [tox]
-envlist = py{36,37,38}-{offline},online,build_docs,figure
-isolated_build = True
+envlist =
+    py{36,37,38}-test{,-oldestdeps,-devdeps,-online}
+    build_docs
+    figure
+    figure_astropydev
+    conda
+    codestyle
+isolated_build = true
 
 [testenv]
+
+# We use bash in some of our enviroments so we have to whitelist it.
 whitelist_externals=
     /bin/bash
     /usr/bin/bash
-changedir = tmp
+
+# Run the tests in a temporary directory to make sure that we don't import
+# astropy from the source tree
+changedir = .tmp/{envname}
+
+# tox environments are constructued with so-called 'factors' (or terms)
+# separated by hyphens, e.g. test-devdeps-cov. Lines below starting with factor:
+# will only take effect if that factor is included in the environment name. To
+# see a list of example environments that can be run, along with a description,
+# run:
+#
+#     tox -l -v
+#
+description =
+    run tests
+    alldeps: with all optional dependencies
+    devdeps: with the latest developer version of key dependencies
+    oldestdeps: with the oldest supported version of key dependencies
+    cov: and test coverage
+
 setenv =
     MPLBACKEND = agg
     COLUMNS = 180
-    PYTEST_COMMAND = pytest -vvv --pyargs sunpy --cov=sunpy --cov-config={toxinidir}/setup.cfg -m "not figure" {toxinidir}/docs
-    astropydev: HOME = {envtmpdir}
-extras = all,tests
+    PYTEST_COMMAND = pytest -vvv --pyargs sunpy --cov=sunpy --cov-config={toxinidir}/setup.cfg {toxinidir}/docs
+    astropydev,build_docs,online: HOME = {envtmpdir}
+
 deps =
-    pytest-cov
+
+    # All our tests run in parallel which this plugin allows us to.
     pytest-xdist
-    astropydev: cython
-    astropydev,figure_astropydev: git+https://github.com/astropy/astropy
-    astropy31: astropy<3.2
+
+    # The devdeps factor is intended to be used to install the latest developer version.
+    # of key dependencies.
+    devdeps: cython
+    devdeps,figure_astropydev: git+https://github.com/astropy/astropy
+    devdeps: astropy<3.2
```

Comment:
wait how are we installing git astropy and astropy <3.2?!

Reply:
Also, we should add some more oldest deps to this, like mpl, numpy whatever else we pin in setup.cfg.

Reply:
Woops.

Reply:
Woops.